core/network/src/protocol/sync/extra_requests.rs

```diff
@@ -14,416 +14,369 @@
-		&mut self.finality_proofs
-	}
-
-	/// Dispatches all possible pending requests to the given peers.
-	pub(crate) fn dispatch(&mut self, peers: &mut HashMap<PeerId, PeerSync<B>>, protocol: &mut dyn Context<B>) {
-		self.justifications.dispatch(peers, protocol);
-		self.finality_proofs.dispatch(peers, protocol);
-	}
-
-	/// Removes any pending extra requests for blocks lower than the
-	/// given best finalized.
-	pub(crate) fn on_block_finalized<F>(
-		&mut self,
-		best_finalized_hash: &B::Hash,
-		best_finalized_number: NumberFor<B>,
-		is_descendent_of: &F,
-	) -> Result<(), fork_tree::Error<ClientError>>
-		where F: Fn(&B::Hash, &B::Hash) -> Result<bool, ClientError>
-	{
-		self.justifications.on_block_finalized(best_finalized_hash, best_finalized_number, is_descendent_of)?;
-		self.finality_proofs.on_block_finalized(best_finalized_hash, best_finalized_number, is_descendent_of)?;
-		Ok(())
-	}
-
-	/// Retry any pending request if a peer disconnected.
-	pub(crate) fn peer_disconnected(&mut self, who: PeerId) {
-		self.justifications.peer_disconnected(&who);
-		self.finality_proofs.peer_disconnected(&who);
-	}
-}
+pub(crate) type ExtraRequest<B> = (<B as BlockT>::Hash, NumberFor<B>);
 
 /// Manages pending block extra data (e.g. justification) requests.
+///
 /// Multiple extras may be requested for competing forks, or for the same branch
 /// at different (increasing) heights. This structure will guarantee that extras
 /// are fetched in-order, and that obsolete changes are pruned (when finalizing a
 /// competing fork).
-pub(crate) struct ExtraRequests<B: BlockT, Essence> {
+#[derive(Debug)]
+pub(crate) struct ExtraRequests<B: BlockT> {
 	tree: ForkTree<B::Hash, NumberFor<B>, ()>,
+	/// requests which have been queued for later processing
 	pending_requests: VecDeque<ExtraRequest<B>>,
-	peer_requests: HashMap<PeerId, ExtraRequest<B>>,
-	previous_requests: HashMap<ExtraRequest<B>, Vec<(PeerId, Instant)>>,
+	/// requests which are currently underway to some peer
+	active_requests: HashMap<PeerId, ExtraRequest<B>>,
+	/// previous requests without response
+	failed_requests: HashMap<ExtraRequest<B>, Vec<(PeerId, Instant)>>,
+	/// successful requests
 	importing_requests: HashSet<ExtraRequest<B>>,
-	essence: Essence,
 }
 
-impl<B: BlockT, Essence: ExtraRequestsEssence<B>> ExtraRequests<B, Essence> {
-	fn new(essence: Essence) -> Self {
+impl<B: BlockT> ExtraRequests<B> {
+	pub(crate) fn new() -> Self {
 		ExtraRequests {
 			tree: ForkTree::new(),
 			pending_requests: VecDeque::new(),
-			peer_requests: HashMap::new(),
-			previous_requests: HashMap::new(),
+			active_requests: HashMap::new(),
+			failed_requests: HashMap::new(),
 			importing_requests: HashSet::new(),
-			essence,
 		}
 	}
 
-	/// Get mutable reference to the requests essence.
-	pub(crate) fn essence(&mut self) -> &mut Essence {
-		&mut self.essence
+	/// Reset all state as if returned from `new`.
+	pub(crate) fn reset(&mut self) {
```

Comment:
Why would someone use this and not `new` instead?

Reply:
Generally speaking for the same reason why one would use the various `clear` methods on collection types―to reuse memory.